MoreLinq/TrySingle.cs

```diff
@@ -0,0 +1,171 @@
+                            theCollection is IList<T> theList ? theList[0]
+                                                              : theCollection.First());
+                    default:
+                        return resultSelector(many, default);
+                }
+            }
+
+            TResult TrySingleForEnumerable(IEnumerable<T> theEnumerable)
+            {
+                using (var e = theEnumerable.GetEnumerator())
+                {
+                    if (!e.MoveNext())
+                        return resultSelector(zero, default);
+                    var current = e.Current;
+                    return !e.MoveNext() ? resultSelector(one, current)
+                                         : resultSelector(many, default);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Attempts to retrieve and project the only element of a sequence,
+        /// similar to <see cref="Enumerable.Single{T}(IEnumerable{T})"/>.
+        /// Unlike that extension method if the sequence contains zero or many
+        /// elements this operator doesn't throw an exception, rather it returns
+        /// a tuple containing the observed cardinality of the sequence and
+        /// either the single value (if only one element in the sequence) or
+        /// the default value of <typeparamref name="T"/>. This enables the
+        /// caller to differentiate between the case where a sequence contains
+        /// zero elements and many elements.</summary>
+        /// <param name="source">
+        /// The source sequence that will be tested for its cardinality.</param>
+        /// <param name="zero">
+        /// The value that is returned in the tuple if the sequence has zero
+        /// elements.</param>
+        /// <param name="one">
+        /// The value that is returned in the tuple if the sequence has a
+        /// single element only.</param>
+        /// <param name="many">
+        /// The value that is returned in the tuple if the sequence has two or
+        /// more elements.</param>
+        /// <typeparam name="T">
+        /// The type of the elements of the sequence.</typeparam>
+        /// <typeparam name="TCardinality">
+        /// The type that expresses cardinality.</typeparam>
+        /// <returns>
+        /// A tuple containing the identified <typeparamref name="TCardinality"/>
+        /// and either the single value of <typeparamref name="T"/> in the sequence
+        /// or its default value.</returns>
+        /// <remarks>
+        /// This operator uses immediate execution, but never consumes more
+        /// than two elements from the sequence. When the source sequence is an
+        /// <see cref="IList{T}"/> or <see cref="ICollection{T}"/> then the
+        /// implementation optimizes by checking the number of elements in
+        /// the underlying sequence.</remarks>
+        public static (TCardinality Cardinality, T Value)
```

Comment:
Add a blank line between doc summary and method declaration for clarity:

```suggestion

        public static (TCardinality Cardinality, T Value)
```